pycqed/instrument_drivers/meta_instrument/qubit_objects/mock_CCL_Transmon.py

```diff
@@ -806,11 +1091,75 @@ def measure_ramsey(self, times=None, MC=None,
                    'frequency': a.qubit_frequency}
             return res
 
+    def measure_echo(self, times=None, MC=None, analyze=True, close_fig=True,
+                     update=True, label: str = ''):
+
+        if MC is None:
+            MC = self.instr_MC.get_instr()
+
+        if times is None:
+            stepsize = (self.T2_echo()*2/61)//(abs(self.cfg_cycle_time())) \
+                * abs(self.cfg_cycle_time())
+            times = np.arange(0, self.T2_echo()*4, stepsize*2)
+
+        dt = times[1] - times[0]
+        times = np.concatenate([times,
+                                (times[-1]+1*dt,
+                                 times[-1]+2*dt,
+                                 times[-1]+3*dt,
+                                 times[-1]+4*dt)])
+
+        s = swf.None_Sweep(parameter_name='Time', unit='s')
+
+        low_lvl = self.measurement_signal(excited=False)
+        high_lvl = self.measurement_signal(excited=True)
+
+        freq_qubit = self.calculate_mock_qubit_frequency()
+
+        detuning = np.abs(self.freq_qubit() - freq_qubit)/1e6
+        highlow = (high_lvl-low_lvl)*np.exp(-detuning)
+
+        high_lvl = low_lvl + highlow
+
+        signal_amp = (high_lvl - low_lvl)/2
+        offset = (high_lvl + low_lvl)/2
+
+        phase = np.pi
+        oscillation_offset = 0
+        exponential_offset = offset
+        frequency = 4/times[-1]  # 4 oscillations
+
+        # Mock values without calibration points
+        mocked_values = (signal_amp *
+                         np.exp(-(times[0:-4] / self.mock_T2_echo())) *
+                         (np.cos(2*np.pi*frequency*times[0:-4] + phase) +
+                          oscillation_offset) + exponential_offset)
+
+        mocked_values = self.values_to_IQ(mocked_values)
+
+        d = det.Mock_Detector(value_names=['raw w1', 'raw w0'],
+                              value_units=['V', 'V'],
+                              detector_control='soft',
+                              mock_values=mocked_values)
+
+        MC.set_sweep_function(s)
+        MC.set_sweep_points(times)
+        MC.set_detector_function(d)
+        MC.run('mock_echo' + self.msmt_suffix)
+
+        if analyze:
+            # N.B. v1.5 analysis
+            a = ma.Echo_analysis_V15(label='echo', auto=True, close_fig=True)
+            if update:
+                self.T2_echo(a.fit_res.params['tau'].value)
+            return a.fit_res.params['tau'].value
+
     def measure_T1(self, times=None, MC=None, analyze=True, close_fig=True,
                    update=True, prepare_for_timedomain=True):
         '''
-        Very simple version that just returns a exponential decay based on mock_T1.
-        Might be improved by making it depend on how close your pulse amp is.
+        Very simple version that just returns a exponential decay based on
+        mock_T1. Might be improved by making it depend on how close your
```

Comment:
Good comment! Makes for a good test case that when you "break" the pulse amp and run this, the T1 no longer works forcing the dep graph to go back and recover from the failure.